File: grafo_nao_direcionado_1.py
class Grafo:
    def __init__(self): #construtor de uma classe em python
        self.grafo = {} #dicionario vazio que sera usado para representar grafo
    
    def adicionar_vertice(self, vertice):
        if vertice not in self.grafo:
            self.grafo[vertice] = []

    def adicionar_aresta(self, vertice1, vertice2):
        #Add uma aresta não direcionada entre vertice1 e vertice2
        if vertice1 in self.grafo:
            self.grafo[vertice1].append(vertice2)
        if vertice2 in self.grafo:
            self.grafo[vertice2].append(vertice1)
    
    def mostrar_grafo(self):
        for vertice in self.grafo:
            print(f"{vertice}: {self.grafo[vertice]}")
    

    #exemplo de uso

grafo = Grafo()

grafo.adicionar_vertice("A")
grafo.adicionar_vertice("B")
grafo.adicionar_vertice("C")
grafo.adicionar_vertice("D")

grafo.adicionar_aresta("A", "B")
grafo.adicionar_aresta("B", "C")
grafo.adicionar_aresta("C", "D")
grafo.adicionar_aresta("D", "A")

grafo.mostrar_grafo()

File: grafo_nao_direcionado_1.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>

using  std::vector;
using  std::string;
using  std::cout;
using  std::endl;
using  std::unordered_map;

class Grafo {

private:
    unordered_map<string, vector<string>> grafo;


    public:
        void adicionarVertice(string vertice){
            if (grafo.find(vertice) == grafo.end()){
                grafo[vertice] = vector<string>();
            }
        }

    

};

/*
- unordered_map : classe contêiner associativo que armazena pares chaves-valor. Neste caso, está sendo usada para criar um mapeamento
entre strings (chaves) e vetores de strings (valores).

- grafo: nome da variável que representa o unordered map

- cada chave no unordered_map representa um vértice no grafo

- unordered_map<string, vector<string>> grafo está declarando uma estrutura de dados que pode ser  representad


*/